src/index/LocatedTriples.cpp

```diff
@@ -0,0 +1,243 @@
+// Copyright 2023 - 2024, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Authors:
+//    2023 Hannah Bast <bast@cs.uni-freiburg.de>
+//    2024 Julian Mundhahs <mundhahj@tf.uni-freiburg.de>
+
+#include "index/LocatedTriples.h"
+
+#include <algorithm>
+
+#include "absl/strings/str_join.h"
+#include "index/CompressedRelation.h"
+#include "index/IndexMetaData.h"
+#include "index/Permutation.h"
+
+// ____________________________________________________________________________
+std::vector<LocatedTriple> LocatedTriple::locateTriplesInPermutation(
+    const std::vector<IdTriple<0>>& triples,
+    const std::vector<CompressedBlockMetadata>& blockMetadata,
+    const std::array<size_t, 3>& keyOrder, bool shouldExist,
+    ad_utility::SharedCancellationHandle cancellationHandle) {
+  vector<LocatedTriple> out;
+  out.reserve(triples.size());
+  size_t currentBlockIndex;
+  for (auto triple : triples) {
+    triple = triple.permute(keyOrder);
+    currentBlockIndex =
+        std::ranges::lower_bound(
+            blockMetadata, triple.toPermutedTriple(),
+            [&](const CompressedBlockMetadata::PermutedTriple& block,
+                const CompressedBlockMetadata::PermutedTriple& triple) {
+              return block < triple;
+            },
+            [](const CompressedBlockMetadata& block) {
+              return block.lastTriple_;
+            }) -
+        blockMetadata.begin();
+    out.emplace_back(currentBlockIndex, triple, shouldExist);
+    cancellationHandle->throwIfCancelled();
+  }
+
+  return out;
+}
+
+// ____________________________________________________________________________
+bool LocatedTriplesPerBlock::hasUpdates(size_t blockIndex) const {
+  return map_.contains(blockIndex);
+}
+
+// ____________________________________________________________________________
+NumAddedAndDeleted LocatedTriplesPerBlock::numTriples(size_t blockIndex) const {
+  // If no located triples for `blockIndex_` exist, there is no entry in `map_`.
+  if (!hasUpdates(blockIndex)) {
+    return {0, 0};
+  }
+
+  auto blockUpdateTriples = map_.at(blockIndex);
+  size_t countInserts = std::ranges::count_if(
+      blockUpdateTriples, &LocatedTriple::shouldTripleExist_);
+  return {countInserts, blockUpdateTriples.size() - countInserts};
+}
+
+// ____________________________________________________________________________
+// Collect the relevant entries of a LocatedTriple into a triple.
+template <size_t numIndexColumns>
+auto tieIdTableRow(auto& row) {
+  if constexpr (numIndexColumns == 3) {
+    return std::tie(row[0], row[1], row[2]);
+  } else if constexpr (numIndexColumns == 2) {
+    return std::tie(row[0], row[1]);
+  } else {
+    return std::tie(row[0]);
+  }
+}
+
+// ____________________________________________________________________________
+// Collect the relevant entries of a LocatedTriple into a triple.
+template <size_t numIndexColumns>
+auto tieLocatedTriple(auto lt) {
+  if constexpr (numIndexColumns == 3) {
```

Comment:
bind `lt-triple_.ids_` to a variable, s.t. the rest of the code becomes more readable.